### DISCUSSION TITLE
Placing Objects Anywhere is now permanently impossible

### DISCUSSION THREAD

**Discussion by winstxnhdw** [winstxnhdw/lc-hax#236] [Announcements]
Bad news everyone. It seems that placing an object is now validated individually on each client. If you try to build outside of the ship, the client will just bring it back into the ship. This means the following features are now permanently broken:

- Build Anywhere
- Teleport Player to Player
- Void
- Random (partially)

It's a real shame since we had an entire ecosystem of tools and utilities dedicated to building which has now (almost) all gone to waste. I reckon I spent more than 50% of development time on the build features since it was what made this project stand out from others. Here's Zeekers's current implementation to validate on the client.

```cs
	public void PlaceShipObject(Vector3 placementPosition, Vector3 placementRotation, PlaceableShipObject placeableObject, bool placementSFX = true)
	{
		Vector3 rotationOffset = placeableObject.parentObject.rotationOffset;
		StartOfRound.Instance.suckingFurnitureOutOfShip = false;
		StartOfRound.Instance.unlockablesList.unlockables[placeableObject.unlockableID].placedPosition = placementPosition;
		StartOfRound.Instance.unlockablesList.unlockables[placeableObject.unlockableID].placedRotation = placementRotation;
		Debug.Log(string.Format("Saving placed position as: {0}", placementPosition));
		StartOfRound.Instance.unlockablesList.unlockables[placeableObject.unlockableID].hasBeenMoved = true;
		if (placeableObject.parentObjectSecondary != null)
		{
			Vector3 vector = placeableObject.parentObjectSecondary.transform.position;
			Quaternion rotation = placeableObject.parentObjectSecondary.transform.rotation;
			Quaternion lhs = Quaternion.Euler(placementRotation) * Quaternion.Inverse(placeableObject.mainMesh.transform.rotation);
			placeableObject.parentObjectSecondary.transform.rotation = lhs * placeableObject.parentObjectSecondary.transform.rotation;
			placeableObject.parentObjectSecondary.position = placementPosition + (placeableObject.parentObjectSecondary.transform.position - placeableObject.mainMesh.transform.position) + (placeableObject.mainMesh.transform.position - placeableObject.placeObjectCollider.transform.position);
			if (!StartOfRound.Instance.shipInnerRoomBounds.bounds.Contains(placeableObject.parentObjectSecondary.transform.position))
			{
				placeableObject.parentObjectSecondary.transform.position = vector;
				placeableObject.parentObjectSecondary.transform.rotation = rotation;
			}
		}
		else if (placeableObject.parentObject != null)
		{
			Vector3 vector = placeableObject.parentObject.positionOffset;
			Quaternion rotation = placeableObject.parentObject.transform.rotation;
			Quaternion lhs2 = Quaternion.Euler(placementRotation) * Quaternion.Inverse(placeableObject.mainMesh.transform.rotation);
			placeableObject.parentObject.rotationOffset = (lhs2 * placeableObject.parentObject.transform.rotation).eulerAngles;
			placeableObject.parentObject.transform.rotation = lhs2 * placeableObject.parentObject.transform.rotation;
			placeableObject.parentObject.positionOffset = StartOfRound.Instance.elevatorTransform.InverseTransformPoint(placementPosition + (placeableObject.parentObject.transform.position - placeableObject.mainMesh.transform.position) + (placeableObject.mainMesh.transform.position - placeableObject.placeObjectCollider.transform.position));
			if (!StartOfRound.Instance.shipInnerRoomBounds.bounds.Contains(placeableObject.parentObject.transform.position))
			{
				placeableObject.parentObject.positionOffset = vector;
				placeableObject.parentObject.transform.rotation = rotation;
				placeableObject.parentObject.rotationOffset = rotationOffset;
			}
		}
		if (placementSFX)
		{
			placeableObject.GetComponent<AudioSource>().PlayOneShot(placeableObject.placeObjectSFX);
		}
	}
```

I blame the low IQ mods that have the dumbass feature to place the terminal far away, ruining the game permanently just for braindead funnies, which galvanised this change.

**Comment by winstxnhdw** [comment]
I am considering to remove `pumpkin` as well.

**Comment by o777l** [comment]
Damn that sucks. Barely got to use these ones properly.

**Comment by joep26020** [comment]
aw rip. i loved trapping people. and i feel bad for you too winston. hard work gone to waste! wondering though  what if you could spoof ship location to place an object in a spot. for example, this terminal moves with the ship. so, what if you put the ship at the entrance only on your client and then put the terminal there. would that work?

**Reply by winstxnhdw** [reply]
You can't because you would have to be able to spoof other people's ship locations as well, which isn't possible. At the very least, it would only appear on client and you could probably use the terminal there, but no one else will be able to see it.

**Comment by sinaxzq** [comment]
hmm, build anywhere function still works in other cheats, or you just used some special way to do it that got blocked? (still not a programmer)

**Reply by winstxnhdw** [reply]
Which other cheat isit still working with? Did you confirm it yourself? Is it on v49? Some mods like Project Apparatus just haven't updated.

**Reply by sinaxzq** [reply]
playing on v49, lethalmenu 1.3.2, did it 30m ago

**Reply by winstxnhdw** [reply]
Are you host or no?

**Reply by sinaxzq** [reply]
nope. and i think it worked with your cheat on v49 before you removed it (/ct too), or did they fix it but didnt change version to v50?

**Reply by winstxnhdw** [reply]
It definitely does not work on this project. In fact, I haven't completely remove it. Try doing `/build` about 30m away from the ship. Shouldn't be possible.

**Reply by sinaxzq** [reply]
try with lethalmenu or lc-hax?

**Reply by winstxnhdw** [reply]
lc-hax

**Reply by sinaxzq** [reply]
ohh it works that way, never used this command. I meant function to be able to go out from the ship with selected item (i mean bunkbeds or terminal)  and set it by yourself
![image](https://github.com/winstxnhdw/lc-hax/assets/127883562/28b97dea-98bd-4b54-a410-c9a6358610d8)

**Reply by sinaxzq** [reply]
thats what i mean

![image](https://github.com/winstxnhdw/lc-hax/assets/127883562/9269b21a-b3ee-4748-9627-5ab78b38fdd2)

**Reply by winstxnhdw** [reply]
You can set it?

**Reply by sinaxzq** [reply]
thats lethalmenu forgot to say, but first picture is your cheat

**Reply by winstxnhdw** [reply]
I am so confused. I'll try it again.

**Reply by winstxnhdw** [reply]
Just tested. Seems like you can only build it in an invalid position once. Afterwards, it’s bugged forever and you can’t move it.

**Reply by sinaxzq** [reply]
oh, yes,  then maybe you need a function as in lethalmenu to be able to replace outside the ship as on second pic

**Reply by winstxnhdw** [reply]
I'll give it a try.

**Reply by sinaxzq** [reply]
oh, sorry, forget, yep you cant replace things when they are set outside, and even if you remove it (with lethalmenu) and use /build, it will spawn where u /build it first time. i thought this was even before v49

**Reply by winstxnhdw** [reply]
Sigh, so building is indeed broken forever :(

**Reply by sinaxzq** [reply]
yes:( all that work... at least anti kick still be gold in this cheat😔 maybe you can boost that slow as hell rocket with my shovels? :) or change time or stop it, if its not host only. And /money is broken idk if you know

**Reply by winstxnhdw** [reply]
Can't change time without host. `/money` is working for me. Check your mods.

**Reply by sinaxzq** [reply]
huh it NEVER worked for me, i mean i got money only locally and when i written some things to buy i got error on terminal, but after u wrote that it works it started working, are u god or sum

**Reply by winstxnhdw** [reply]
probably u had some mod that caused problems

**Reply by sinaxzq** [reply]
no no, nothing additional, nor mods nor cheats, forget bout that

**Reply by joep26020** [reply]
how do u get lethal menu

**Reply by sinaxzq** [reply]
https://www.unknowncheats.me/forum/other-fps-games/615575-lethal-menu-lethal-company-cheat.html

**Comment by winstxnhdw** [comment]
We are so back #550
